### ISSUE TITLE
Workflow-level checkpointing

### ISSUE THREAD

**Issue by kylechard** [Parsl/parsl#43]
We need to provide the ability to record workflow state (e.g., task status, input/output, task command) and export it to an external store (e.g., database, file). Allow for restart from the checkpoint.

**Comment by yadudoc** [comment]
Breaking this issue into concrete tasks : 

- [ ] Enable checkpointing functionality in the dfk.
- [ ] Enable option to auto-checkpoint at defined time-intervals and events such as cleanup
- [ ] Restarts from checkpoints. Restarting could be complex for non-deterministic applications.

For workflows with no-nondeterministic pieces such as conditionals or loops on non-deterministic outputs, can be quite hard and the best we can do is, use the checkpoint as a memoization lookup table.
This approach would work correctly for deterministic workflows. However it might be expensive to track, input file timestamps, hashes of function body, etc to know if a the task and its inputs has not changed.

Alternatively, for purely deterministic workflows we could use task_ids (which are assigned sequentially) to track what apps have completed, and allow for restarting from the checkpoint.

**Comment by salmanhabib** [comment]
Also, don't forget having the run info in a database (this part of both the provenance tracking and checkpointing requirements).

**Comment by yadudoc** [comment]
Adding couple more task to track requirements from @salmanhabib  : 

- [ ]  Checkpoint periodically to a database 
- [ ]  Streaming state to a database

**Comment by danielskatz** [comment]
@salmanhabib - can you please define what you mean by "provenance tracking" here?  What specific information do you want to be stored?

**Comment by salmanhabib** [comment]
@danielskatz Things like run parameters, status of outputs, versions of software/compute modules, lists of failed tasks (if relevant), etc. Basically we want to have the notion of reproducibility baked into the workflow so we can rerun things and get the same answer plus compare against runs where we know certain changes/modifications have been made.

**Comment by kylechard** [comment]
@salmanhabib We also have an issue related to capturing broad provenance information: https://github.com/Parsl/parsl/issues/42.

**Comment by salmanhabib** [comment]
@kylechard Yes, I know, originally I was just saying that there is some overlap between the two.

**Comment by yadudoc** [comment]
Here's what I propose : 

The DFK will have two methods:
1. checkpoint( ) - write all non-checkpointed info to a checkpoint object
2. resume_from( ) that loads all checkpointed information from a checkpoint object

We will support three modes for automatic checkpointing:
1.  (eager) checkpoint at the completion of every task or 
2.  (lazy) at configurable intervals (say ~1hr).
3.  (at_exit) write out a checkpoint when the DFK exits, capturing early termination from app failures,
uncaught exceptions, keyboard interrupts etc.

In the simplest case, code does not change, functions don't hold any state (no clojures) and are deterministic. So if an app executes a function named app_foo( ), every subsequent invocation with the same args, kwargs and env would produce the same results. With this assumption, if we store completed results and return these results at call time as completed futures we would skip every invocation of a task that has been completed as we re-create the dataflow-graph on re-execution.

Complications: 
* Tracking code change, and forcing re-execution of parts of the graph.
* Function could be redefined in different contexts
* Data objects could be very deep or not comparable
* Changes to file objects aren't trackable since they are right now treated as strings, this would change once File objects are used consistently, but would require means to compare files.
* Non-determinism from the script itself via conditions

**Comment by annawoodard** [comment]
This is really exciting!

Just brainstorming: could we tweak `resume_from` to be more general? One possibility would be to have a `load_cache` or similar so checkpoints could be composed. For example, if you have two workflows, `blue` and `red`, that look like this:

```
blue: 
    A -> B -> C
    D -> E
    C, E -> F
red:
    F -> G -> H
yellow:
    D -> E
    F -> G
    E, G -> I
```

Say you've already run the blue and red workflows. Now you want to run the yellow one. With the composing feature, you could use: `load_cache(blue, red)` and re-use *both* the `D -> E` and `F -> G` steps.

**Comment by yadudoc** [comment]
This should be doable since resume_from(...) should be capable of loading multiple checkpoint files. The yellow workflow has to be a mashup of parts of both the blue and red workflow to use results from either since the results are made available only when the parent apps are called.

Some other things to consider:
* checkpointing is closely related to provenance tracking and monitoring, and it might help to design the interface keeping future deliverables like pushing state info to a DB
* do we include a previous checkpoint in a run that loads from a checkpoint ?

**Comment by annawoodard** [comment]
> This should be doable since resume_from(...) should be capable of loading multiple checkpoint files. 

Right, "loading multiple checkpoint files" is the functionality I was advocating for, I'm glad to hear it's already part of the plan. This is minor, but in my opinion `load_cache(red, blue)` or `load_checkpoints(A, B)` would more clearly describe what's being done-- `resume_from` for me implies something being done in serial, it's not clear what `resume_from(red, blue)` means (are we starting from red or blue?).

> The yellow workflow has to be a mashup of parts of both the blue and red workflow to use results from either since the results are made available only when the parent apps are called.

Yes, exactly. In my work, it's a common scenario. Checkpointing "red workflow" so that "red workflow" can be restarted without waste is exciting. Checkpointing "red workflow" so work done can be re-used in subsequent workflows is still more exciting! In HEP this is commonly achieved "by hand" by splitting work into stages depending on how often you're likely to make changes to each stage. That way the less-frequently-changing stages don't have to be re-run. Parsl supporting multiple checkpoint files would abstract that "by hand" factorization away.

> checkpointing is closely related to provenance tracking and monitoring, and it might help to design the interface keeping future deliverables like pushing state info to a DB

I was assuming that the checkpoint object would stored in a DB. For optimal speed I think that's the way to go. What is the current proposal?

> do we include a previous checkpoint in a run that loads from a checkpoint ?

This is an important question and a bit tricky. The more unneeded "stuff" that's being carried around in the checkpoints, the slower/less convenient they will be. I think we would definitely want to at least have the option to "prune" a checkpoint such that anything from the previous checkpoint that is not used in the current workflow is dropped.

**Comment by yadudoc** [comment]
We can now do these things : 
* Memoize apps 
* Memoization to use code changes (works in interactive envs as well)
* Write checkpoint incrementally (Manual, user calls dfk.checkpoint)
* Load checkpoints at start time.
* Dev docs added, user docs pending.

**Comment by yadudoc** [comment]
Merged to master as of 99e9461a728039ae6c0b86ab11204922e35df73e
